### ISSUE TITLE
fix: ghash->uint64 bitcast_convert_type miscompiles on the CPU PJRT backend

### ISSUE THREAD

**Issue by chokobole** [fractalyze/flock-zorch#75]
### Pre-submission checklist

- [x] I have searched the existing issues and this is not a duplicate.

### Affected version / commit

flock-zorch 2bdd90b; jax / jaxlib / jax-cuda12-plugin / jax-cuda12-pjrt pin 0.10.0.dev20260707115835

### Environment

Linux

### Reproducer

```shell
# On the CPU PJRT backend the direct ghash->uint64 bitcast returns wrong lanes;
# the uint32-routed path (what field.from_ghash does) is correct. On GPU both agree.
JAX_PLATFORMS=cpu .venv/bin/python - <<'PY'
import numpy as np, jax, jax.numpy as jnp
from jax import lax
jax.config.update("jax_enable_x64", True)
import zk_dtypes
g  = jnp.binary_field_ghash
u  = jnp.asarray([[5, 7]], jnp.uint64)                    # one F128 = {lo: 5, hi: 7}
gh = lax.bitcast_convert_type(u, g)                       # uint64 -> ghash (correct on CPU)
direct = np.asarray(lax.bitcast_convert_type(gh, jnp.uint64))                     # ghash -> uint64, DIRECT
u32    = lax.bitcast_convert_type(gh, jnp.uint32)                                 # ghash -> uint64, ROUTED
routed = np.asarray(lax.bitcast_convert_type(u32.reshape(*u32.shape[:-1], 2, 2), jnp.uint64))
print("direct:", direct.tolist(), " routed:", routed.tolist(), " want: [[5, 7]]")
PY
```

### Expected behavior

`ghash -> uint64` `lax.bitcast_convert_type` returns the original {lo, hi} uint64 lanes on every backend, matching the uint32-routed path.

### Actual behavior

On the CPU PJRT backend the direct `ghash -> uint64` bitcast returns wrong lanes (the round-trip fails); the uint32-routed path returns `[[5, 7]]`. On GPU both are correct. Probed batched (1024) nonzero, eager and jit.

This is why `field.from_ghash` keeps a uint32-lane detour — a raw inline `ghash -> uint64` bitcast would pass a GPU spot-check yet silently corrupt the CPU gate backend (`bazel test //python:all` runs `JAX_PLATFORMS=cpu`).

The *reverse* direction, `uint64 -> ghash`, is now correct on CPU, so only `ghash -> uint64` remains broken — this refines #62, whose premise was that the `uint64 -> ghash` direction miscompiles on CPU. Root cause is the fractalyze/xla `BitcastConvertType` lowering on the CPU PJRT path; zorch#399 tracked/fixed the GPU side (where `ghash -> uint64` returned zeros).

flock-zorch action: keep the `field.from_ghash` routing; drop it once the CPU lowering is fixed and re-verify the CPU gates.

### Relevant logs / stack trace

```text
CPU backend (JAX_PLATFORMS=cpu), batched 1024 nonzero:
[eager] to: direct==routed=True  gt=True | from: direct==routed=False  direct_roundtrip=False
[jit]   to: direct==routed=True  gt=True | from: direct==routed=False  direct_roundtrip=False
GPU backend: all True (both directions correct).
```

### Area

compiler

### Milestone

zorch-native refactor

### Relevant knowledge

`uint64[...,2] <-> binary_field_ghash` bitcast on the current jax pin (2026-07-08):

| direction        | CPU direct | GPU direct |
|------------------|------------|------------|
| uint64 -> ghash  | correct    | correct    |
| ghash  -> uint64 | **wrong**  | correct    |

The gates run `JAX_PLATFORMS=cpu`, so the `ghash -> uint64` bitcast must stay routed through uint32 lanes (`field.from_ghash`) or read host bytes — never inline it. `convert_element_type` / `jnp.asarray(dtype=ghash)` are value-casts, not bitcasts (the latter SIGSEGVs).

**Comment by chokobole** [comment]
The workaround this issue documents is now consolidated in one place by #77 (merged): the `ghash -> uint64` direction lives only in `field.from_ghash` (device, uint32-routed) and `field.from_ghash_host` (host np); no consumer inlines a raw `ghash -> uint64` bitcast anymore. So when the upstream CPU `BitcastConvertType` lowering is fixed, dropping the detour is a change to those two `field` functions alone — then re-run `bazel test //python:all` (it runs `JAX_PLATFORMS=cpu`, the backend that currently miscompiles).

**Comment by chokobole** [comment]
### Root cause found + fixed upstream (XLA), verified end-to-end

The CPU `ghash -> uint64` miscompile is an **undefined-behavior bug in XLA's `BitcastDtypesExpander`**, not anything flock-specific.

`xla/hlo/transforms/expanders/bitcast_dtypes_expander.cc` built the narrowing-path low-bits mask as `(int64_t{1} << output_bit_width) - 1`. When `output_bit_width == 64` the shift count equals the operand width — UB; on x86 the count is taken mod 64, so `1 << 64 == 1` and the mask collapses to `0`. The following `And` then zeroes the result.

Confirmed directly in the optimized CPU HLO — the mask constant is literally `u128[] constant("0x0")`, vs `0xffffffff` on the working `-> uint32` path (`output_bit_width == 32`, no UB). That exactly explains the asymmetry: `-> uint32` ✓, `-> uint64` ✗, `uint64 -> ghash` ✓ (the last takes the *widening* branch, which has no mask). GPU is unaffected because it lowers these bitcasts through the native field emitter, not this expander (that path was fractalyze/zorch#399's GPU fix).

**Fix:** fractalyze/xla#241 — build the mask with a well-defined unsigned shift, plus a `U128 -> U64` expander unit test pinning the mask to `0xffffffffffffffff` (the pre-fix code emitted `0x0`).

**Verified end-to-end:** built a CPU jaxlib from #241 (XLA at the jax-pinned `651d22fe` + the fix) and ran the reproducer on `JAX_PLATFORMS=cpu`:
- direct `ghash -> uint64` now `[[5, 7]]` (was `[[0, 0]]`)
- zorch#399's `[[2,0],[3,0],[0,1]]` reproducer ✓
- batched 1024 round-trip, eager + jit ✓

**Remaining (pin-gated) — the drop this issue tracks:**
1. Merge fractalyze/xla#241.
2. Publish a jaxlib to the fractalyze index that includes it.
3. Bump flock-zorch's jax/jaxlib pin (`requirements.in` + lock, mirroring zorch).
4. Then drop the `field.to_ghash`/`from_ghash` uint32 detour (both directions are correct on the fixed backend) and re-run `bazel test //python:all` (`JAX_PLATFORMS=cpu`).

Keeping the routing in place until step 3 lands — the bazel gate resolves jaxlib from the pinned index wheel, which still has the bug.

**Comment by chokobole** [comment]
Prepared the detour-drop as draft PR #82 (`refactor/drop-ghash-uint32-detour-75`) — both `to_ghash`/`from_ghash` collapse to direct bitcasts, local-verified against a CPU jaxlib built from fractalyze/xla#241 (round-trip, batched, host path, dtype-storage bytes all match).

Re-marking `status:blocked`: the drop unblocks once xla#241 ships in a published jaxlib and the pin bumps (4-step recipe above). PR #82 stays draft with expected-red CI until then; merging it closes this issue.

**Comment by chokobole** [comment]
### Detour-drop is unblocked but coupled to a jax-stack bump — plan below

The bitcast fix (fractalyze/xla#241) is merged and lands in jaxlib `dev20260708221109` (pins XLA `49e9376f` = the #241 merge). Dropping the detour requires flock to bump to that wheel, and that bump surfaced two coupling issues, both now understood + fixed locally (full `bazel test //python:all` green, 21/21):

1. **Dual jaxlib hubs.** flock's tests load jaxlib from `flock_zorch_pip` **and** `zorch_pip` (via the zorch module dep), and zorch's hub wins. So flock's pin must move **in lockstep with zorch's** — the `requirements.in` comment was right.

2. **`zorch.sha256` composite went rank-strict.** `dev20260708221109` wires the Sha256Fusion emitter (fractalyze/xla#232/#235); it demands `u32[B,nblocks,16]` and rejects the 4-D operand that zorch's `reconstruct_roots` produced by `vmap`-ing the per-leaf hash+compress. The old wheel *inlined* the composite (vmap-safe). Fixed in **fractalyze/zorch#430** (`reconstruct_roots` batches over leaves / scans depth instead of vmap; byte-identical).

**Remaining steps (in order):**
1. Merge **fractalyze/zorch#430** into `flock-byte-fiat-shamir`.
2. Bump that branch's jax pin to `dev20260708221109` (its usual merge-of-main).
3. flock-zorch: `git_override` → that new zorch HEAD, `requirements.in`/lock → `dev20260708221109`, drop the `field.to_ghash`/`from_ghash` uint32 detour. Then `bazel test //python:all` is green — verified locally with all three in place.

Draft PR #82 will carry step 3 (currently holds a stale detour-drop; will be refreshed once the zorch branch HEAD to pin is decided).

**Comment by chokobole** [comment]
**Correction to the plan** — the sha256/vmap piece moved from a zorch change to an XLA fix, per review.

The zorch `reconstruct_roots` batching (zorch#430) was the wrong layer: reviewer flagged that the `zorch.sha256` *fusion* is the outlier (poseidon2 already handles vmapped markers). Fixed there instead — **fractalyze/xla#243** makes `ZorchFusedRegionRewriter` collapse a vmapped marker's leading dims to the canonical `[B', nblocks, 16]` (emitters unchanged, covers CPU+GPU). zorch#430 is closed.

So the detour-drop no longer needs any zorch source change — just a jaxlib that carries **both** xla#241 (the bitcast fix) and **xla#243** (the sha256/vmap fix). Updated remaining steps:
1. Merge xla#241 (done) + **xla#243**.
2. A published jaxlib includes both → bump flock's jax/jaxlib pin to it.
3. Drop the `field.py` uint32 detour; `bazel test //python:all` green.

(Verified locally that xla#243's rank-4 path byte-matches NIST, and — separately — that the bumped wheel + the reverted zorch batching made the full flock gate green, so both fixes land the same result.)

**Comment by chokobole** [comment]
### Session 2026-07-09 handoff — both XLA fixes merged; detour-drop unblocked

- **Done:**
  - **fractalyze/xla#241** (the `int64_t{1} << 64` bitcast UB) merged (`49e9376f`).
  - **fractalyze/xla#243** (sha256 vmap — the fusion rewriter collapses `rank>3` blocks to canonical `[B',nblocks,16]`, mirroring poseidon2; emitters unchanged, CPU+GPU) merged (`30e1d735`). **fractalyze/zorch#430 closed** — superseded, since the sha256 fix is jaxlib-level and needs no zorch source change.
  - Verified jaxlib **`dev20260709073514`** (jax@`336c49fd`, XLA `bcc2a928`) is an ancestor-check ✓ for **both** fixes.

- **Next session — bump the pins and drop the detour:**
  1. The zorch **`flock-byte-fiat-shamir`** branch (flock's `git_override` in `MODULE.bazel`) must pin `dev20260709073514` (its normal merge-of-main; no source change). flock's `//python:all` tests load jaxlib from **both** `flock_zorch_pip` *and* `zorch_pip` (via the zorch dep), and the zorch hub wins the import — so *both* pins must carry the fixed wheel, else the old buggy jaxlib is used.
  2. flock-zorch: `MODULE.bazel` `git_override` → that zorch commit; `requirements.in` + regen `requirements_lock_3_11.txt` → `dev20260709073514` (`bazel run //:requirements.update`).
  3. Drop the detour in `python/flock_zorch/field/f128.py`: `to_ghash`/`from_ghash` become single direct `jax.lax.bitcast_convert_type` calls (`uint64[...,2]` ⇄ `binary_field_ghash`); trim the module + fn docstrings that describe the uint32 routing.
  4. `bazel test //python:all` (runs `JAX_PLATFORMS=cpu`) → green → `/create-pr` with `Closes #75` (refresh draft PR #82, which currently holds an older detour-drop on the pre-`field/`-subpackage path).

- **Decision tree if the gate is red:**
  - `*_oracle_test` shows `got=[0, 0]` → the jaxlib in use lacks xla#241; confirm **both** pip hubs resolve `dev20260709073514+`.
  - `zorch.sha256 composite ... got u32[…,…,…,16]` rank error → lacks xla#243; same both-hubs pin check.

- **Reference files:** `python/flock_zorch/field/f128.py` (detour), `requirements.in` / `requirements_lock_3_11.txt` (flock pin), `MODULE.bazel` (zorch `git_override`), draft PR fractalyze/flock-zorch#82.

- **Done criteria:** (1) flock + `flock-byte-fiat-shamir` both pin `dev20260709073514+`; (2) `field/f128.py` detour removed; (3) `bazel test //python:all` green; (4) PR merged → #75 closed.

- **Blocker: none** — the wheel with both fixes exists.

- **Already verified locally this session** (both routes land green): (a) the bumped wheel + reverting the zorch batching made the full flock gate pass 21/21; (b) xla#243's rank-4 path byte-matches NIST on CPU and GPU. So the detour-drop lands green once both pins carry the fixes.
